File: sdk/golang/keyforge.go
// Package keyforge provides client SDK capabilities for Go applications.
package keyforge

import (
	"bytes"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strings"
	"time"
)

// Client handles online and offline license validation for Go applications.
type Client struct {
	ProductID     string
	PublicKey     ed25519.PublicKey
	ServerURL     string
	ClientVersion string
	HTTPClient    *http.Client
}

// NewClient initializes a KeyForge client instance.
func NewClient(productID string, serverURL string, pubKey ed25519.PublicKey) *Client {
	return &Client{
		ProductID:     productID,
		PublicKey:     pubKey,
		ServerURL:     strings.TrimRight(serverURL, "/"),
		ClientVersion: "1.0.0",
		HTTPClient:    &http.Client{Timeout: 10 * time.Second},
	}
}

// LicensePayload represents the claims embedded in a license.
type LicensePayload struct {
	SchemaVersion int                    `json:"schema_version"`
	LicenseID     string                 `json:"license_id"`
	LicenseKey    string                 `json:"license_key"`
	ProductID     string                 `json:"product_id"`
	LicenseType   string                 `json:"license_type"`
	Edition       string                 `json:"edition"`
	CustomerID    string                 `json:"customer_id"`
	Features      []string               `json:"features"`
	ExpiresAt     string                 `json:"expires_at"`
	MaxDevices    int                    `json:"max_devices"`
	Metadata      map[string]interface{} `json:"metadata"`
}

// ParsedToken represents an unpacked armored token.
type ParsedToken struct {
	SchemaVersion int
	KeyID         string
	Algorithm     string
	Payload       LicensePayload
	Signature     []byte
}

// ValidationResult represents the output of a validation check.
type ValidationResult struct {
	IsValid       bool     `json:"is_valid"`
	Status        string   `json:"status"`
	Message       string   `json:"message"`
	LicenseID     string   `json:"license_id"`
	ProductID     string   `json:"product_id"`
	Edition       string   `json:"edition"`
	Features      []string `json:"features"`
	ExpiresAt     string   `json:"expires_at"`
	DaysRemaining *int     `json:"days_remaining"`
}

// HasFeature returns true if the license entitles the caller to the named feature.
func (v *ValidationResult) HasFeature(featureName string) bool {
	if !v.IsValid {
		return false
	}
	for _, f := range v.Features {
		if f == "*" || f == featureName {
			return true
		}
	}
	return false
}

// ParseArmoredToken parses a kf1.payload.sig.keyid string.
func ParseArmoredToken(token string) (*ParsedToken, error) {
	parts := strings.Split(strings.TrimSpace(token), ".")
	if len(parts) != 4 || parts[0] != "kf1" {
		return nil, errors.New("invalid KeyForge armored token structure")
	}

	payloadBytes, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil, fmt.Errorf("failed to decode payload: %w", err)
	}

	var payload LicensePayload
	if err := json.Unmarshal(payloadBytes, &payload); err != nil {
		return nil, fmt.Errorf("failed to parse payload JSON: %w", err)
	}

	sigBytes, err := base64.RawURLEncoding.DecodeString(parts[2])
	if err != nil {
		return nil, fmt.Errorf("failed to decode signature: %w", err)
	}

	keyIDBytes, err := base64.RawURLEncoding.DecodeString(parts[3])
	if err != nil {
		return nil, fmt.Errorf("failed to decode key ID: %w", err)
	}

	return &ParsedToken{
		SchemaVersion: 1,
		KeyID:         string(keyIDBytes),
		Algorithm:     "Ed25519",
		Payload:       payload,
		Signature:     sigBytes,
	}, nil
}

// ValidateOnline queries the KeyForge REST endpoint.
func (c *Client) ValidateOnline(licenseKey, installationID string) (*ValidationResult, error) {
	if c.ServerURL == "" {
		return nil, errors.New("server URL required for online validation")
	}

	payload := map[string]string{
		"license_key":     licenseKey,
		"product_id":      c.ProductID,
		"installation_id": installationID,
		"client_version":  c.ClientVersion,
	}
	body, _ := json.Marshal(payload)

	resp, err := c.HTTPClient.Post(fmt.Sprintf("%s/api/v1/licenses/validate", c.ServerURL), "application/json", bytes.NewReader(body))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result ValidationResult
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result, nil
}

File: sdk/lua/keyforge.lua
--[[
    KeyForge Universal Licensing Client Module for Lua
    Compatible with Lua 5.1+, LuaJIT, and OpenResty.
    Provides online REST validation and offline token claim extraction.
--]]

local KeyForge = {}
KeyForge.__index = KeyForge

-- Helper: Base64URL decode
local function b64url_decode(str)
    str = str:gsub("-", "+"):gsub("_", "/")
    local pad = #str % 4
    if pad == 2 then str = str .. "=="
    elseif pad == 3 then str = str .. "=" end
    
    local b = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/'
    return (str:gsub('.', function(x)
        if (x == '=') then return '' end
        local r, f = '', (b:find(x) - 1)
        for i = 6, 1, -1 do r = r .. (f % 2 ^ i - f % 2 ^ (i - 1) > 0 and '1' or '0') end
        return r
    end):gsub('%d%d%d%d%d%d%d%d', function(x)
        if (#x ~= 8) then return '' end
        local c = 0
        for i = 1, 8 do c = c + (x:sub(i, i) == '1' and 2 ^ (8 - i) or 0) end
        return string.char(c)
    end))
end

-- Helper: Basic JSON parse (fallback if cjson not installed)
local function parse_json(str)
    local ok, cjson = pcall(require, "cjson")
    if ok then
        return cjson.decode(str)
    end
    -- Lightweight fallback parser for key-value fields
    local res = {}
    for k, v in string.gmatch(str, '"(%w+)":%s*"([^"]+)"') do
        res[k] = v
    end
    for k, v in string.gmatch(str, '"(%w+)":%s*(%d+)') do
        res[k] = tonumber(v)
    end
    for k, v in string.gmatch(str, '"(%w+)":%s*(true)') do
        res[k] = true
    end
    for k, v in string.gmatch(str, '"(%w+)":%s*(false)') do
        res[k] = false
    end
    return res
end

function KeyForge.new(config)
    local self = setmetatable({}, KeyForge)
    self.product_id = config.product_id or "default-product"
    self.server_url = config.server_url and config.server_url:gsub("/+$", "") or nil
    self.client_version = config.client_version or "1.0.0"
    self.cached_license = nil
    return self
end

-- Parse armored token (kf1.payload.sig.keyid)
function KeyForge:parse_token(token)
    local parts = {}
    for part in string.gmatch(token, "[^.]+") do
        table.insert(parts, part)
    end
    
    if #parts ~= 4 or parts[1] ~= "kf1" then
        return nil, "Invalid KeyForge token structure"
    end
    
    local payload_json = b64url_decode(parts[2])
    local payload = parse_json(payload_json)
    local key_id = b64url_decode(parts[4])
    
    return {
        schema_version = 1,
        key_id = key_id,
        algorithm = "Ed25519",
        payload = payload,
        signature = parts[3]
    }
end

-- Validate license token or key
function KeyForge:validate(license_input)
    if type(license_input) == "string" and license_input:sub(1, 4) == "kf1." then
        local token_obj, err = self:parse_token(license_input)
        if not token_obj then
            return { is_valid = false, status = "INVALID_TOKEN", message = err }
        end
        
        local p = token_obj.payload
        if p.product_id and p.product_id ~= self.product_id then
            return {
                is_valid = false,
                status = "PRODUCT_MISMATCH",
                message = string.format("Product mismatch: expected %s, got %s", self.product_id, p.product_id)
            }
        end
        
        self.cached_license = p
        return {
            is_valid = true,
            status = "VALID",
            license_id = p.license_id,
            product_id = p.product_id,
            edition = p.edition or "standard",
            features = p.features or {},
            expires_at = p.expires_at
        }
    end
    
    return {
        is_valid = false,
        status = "REQUIRES_ONLINE_VALIDATION",
        message = "Raw license key provided without server connection"
    }
end

function KeyForge:has_feature(feature_name)
    if not self.cached_license then return false end
    local feats = self.cached_license.features or {}
    for _, f in ipairs(feats) do
        if f == "*" or f == feature_name then
            return true
        end
    end
    return false
end

return KeyForge
